### PR TITLE
ISPN-15104 Global RemoteCacheManager for remote store

### PR DESCRIPTION
**Pull request by jabolina**
https://issues.redhat.com/browse/ISPN-15104

* Includes the new configuration to create a remote-cache-container
* I am using the marshaller from the first RCS that requests the RCM. If multiple RCS with different marshallers point to the same remote-cache-container could create an issue. I added a verification to limit here, and only create if the marshallers are the same.

### CODE REVIEW

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      if (rcm != null) rcm.stop();
+   }
+
+   @Override
+   public CompletionStage<RemoteCacheManager> cacheContainer(String name, Marshaller marshaller) {
+      CompletionStage<RemoteCacheManager> cs = remoteContainers.computeIfPresent(name, (k, cf) ->
+            cf.thenCompose(prev -> prev != null
+                  ? CompletableFuture.completedFuture(prev)
+                  : createCacheManager(k, marshaller)));
+      // The entry always exists.
```
Comment (wburns) [inline persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java:63]: Is this true? Can't the RemoteStore have the wrong name configured? I think we should check for null and log an ERROR if so.

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/configuration/global/RemoteContainerConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * @since 15.0
+ **/
+public class RemoteContainerConfiguration extends ConfigurationElement<RemoteContainerConfiguration> {
+   static final AttributeDefinition<String> NAME = AttributeDefinition.builder(Attribute.NAME, "", String.class).immutable().build();
+   static final AttributeDefinition<String> URI = AttributeDefinition.builder(Attribute.URI, null, String.class).immutable().build();
+   private final org.infinispan.commons.configuration.attributes.Attribute<TypedProperties> properties;
+
+   protected RemoteContainerConfiguration(AttributeSet attributes, ConfigurationElement<?>... children) {
+      super(Element.REMOTE_CACHE_CONTAINER, attributes, children);
+      this.properties = attributes.attribute(PROPERTIES);
```
Comment (wburns) [inline persistence/remote/src/main/java/org/infinispan/persistence/remote/configuration/global/RemoteContainerConfiguration.java:23]: Are we actually using properties here? Seems unneeded, no?

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/RemoteStore.java [hunk 3]
```diff
@@ -122,19 +123,28 @@ public CompletionStage<Void> start(InitializationContext ctx) {
...
+               .getGlobalComponentRegistry()
+               .getComponent(GlobalRemoteContainers.class);
+         rcmStage = containers.cacheContainer(configuration.remoteCacheContainer(), marshaller);
+      } else {
+         rcmStage = blockingManager.supplyBlocking(() -> {
+            ConfigurationBuilder builder = buildRemoteConfiguration(configuration, marshaller);
+            return new RemoteCacheManager(builder.build());
+         }, "RemoteCacheManager-create");
+      }
+      return rcmStage.thenApplyAsync(rcm -> {
```
Comment (wburns) [inline persistence/remote/src/main/java/org/infinispan/persistence/remote/RemoteStore.java:139]: This should use `BlockingManager.thenApplyBlocking` so we aren't leaking out a blocking thread.

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java [hunk 4]
```diff
@@ -29,27 +33,57 @@ public class GlobalRemoteContainersImpl implements GlobalRemoteContainers {
...
-            remoteContainers.put(e.getKey(), blockingManager.supplyBlocking(() -> new RemoteCacheManager(builder.build()), "RemoteContainer-create"));
+            remoteContainers.put(e.getKey(), CompletableFutures.completedNull());
          }
       }
    }
 
    @Stop
    public void stop() {
-      remoteContainers.values().forEach(c -> c.thenAccept(RemoteCacheManager::stop));
+      remoteContainers.values().forEach(c -> c.thenAccept(this::stop));
```
Comment (wburns) [inline persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java:50]: If we remove the entries from the container and then invoke stop then this component can be restarted if necessary.

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/LifecycleCallbacks.java [hunk 5]
```diff
@@ -22,6 +24,8 @@ public class LifecycleCallbacks implements ModuleLifecycle {
 
    @Override
    public void cacheManagerStarting(GlobalComponentRegistry gcr, GlobalConfiguration globalCfg) {
+      BasicComponentRegistry bcr = gcr.getComponent(BasicComponentRegistry.class);
+      bcr.getComponent(GlobalRemoteContainers.class).running();
```
[the remark this answers is not in the document]
Comment (jabolina) [reply persistence/remote/src/main/java/org/infinispan/persistence/remote/LifecycleCallbacks.java]: Updated. I left only the minimum necessary for the generator to wire the factory and everything. But now is the RCS that invokes `running` to start the component.

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java (hunk 1 above)
Comment (jabolina) [reply persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java]: Updated.

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java (hunk 4 above)
Comment (jabolina) [reply persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java]: Updated. After the close, there is still an entry on the map, but the CF returns null. If the container is requested again, then a new RCM is created.

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+      if (configuration != null) {
+         for (Map.Entry<String, RemoteContainerConfiguration> e : configuration.configurations().entrySet()) {
+            remoteContainers.put(e.getKey(), CompletableFutures.completedNull());
+         }
+      }
+   }
+
+   @Stop
+   public void stop() {
+      remoteContainers.values().forEach(c -> c.thenApply(this::stop));
```
Comment (wburns) [inline persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java:50]: This doesn't replace the value btw, forEach is just a Consumer.

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/configuration/global/RemoteContainerConfiguration.java (hunk 2 above)
Comment (wburns) [reply persistence/remote/src/main/java/org/infinispan/persistence/remote/configuration/global/RemoteContainerConfiguration.java:23]: Were you able to confirm if the properties here are required?

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java (hunk 6 above)
Comment (jabolina) [reply persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java:52]: Updated. Once the reference count reaches zero, the RCM will update the map entry to a CF returning null.

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/configuration/global/RemoteContainerConfiguration.java (hunk 2 above)
Comment (jabolina) [reply persistence/remote/src/main/java/org/infinispan/persistence/remote/configuration/global/RemoteContainerConfiguration.java:23]: I tried allowing the same elements the `remote-store` accepts, but that would be a lot of duplication or changing the existing configuration. As is, we can't use the same builders for the other elements, as they are coupled with the store configuration.

I think we could update it to accept all the same configuration options in other PR/JIRA.

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+   private CompletionStage<RemoteCacheManager> createCacheManager(String name, Marshaller marshaller) {
+      RemoteContainersConfiguration configuration = globalConfiguration.module(RemoteContainersConfiguration.class);
+      if (configuration != null) {
+         RemoteContainerConfiguration rcc = configuration.configurations().get(name);
+
+         if (rcc == null)
+            return CompletableFuture.failedFuture(new IllegalStateException("No configuration defined for container " + name));
+
+         ConfigurationBuilder builder = new ConfigurationBuilder();
+         builder.withProperties(rcc.properties());
```
Comment (wburns) [inline persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java:92]: We should make sure to add some tests to make sure these allow same configuration options as the cache store, especially security based ones.

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+      @Override
+      public void stop() {
+         if (references.decrementAndGet() == 0) {
+            super.stop();
+            remoteCacheManagerShutdown(name);
+         }
+      }
+
+      public RemoteCacheManager increment() {
```
[the remark this answers is not in the document]
Comment (jabolina) [reply persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java]: Updated.

File: persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java (hunk 7 above)
Comment (jabolina) [reply persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java]: Updated.

**Comment by jabolina** [comment]
Updated with all the suggestions. Included a few tests that extend `RemoteStoreFunctionalTest` and for the Hot Rod upgrade tests. The test also uses the properties to configure SSL.

Do we need more updates on the docs side?

**Comment by tristantarrant** [comment]
Merged, thanks

**Pull request merged**